### PR TITLE
Lowering for the Transpose intrinsic

### PR DESCRIPTION
**Pull request by kiranchandramohan**
Note:
1. Created a mutable fir.box of rank 2 for the result matrix. Not sure whether this is the right approach.
2. Does the cleanup handle all cases?
3. Please suggest additional tests.

### CODE REVIEW

File: flang/include/flang/Lower/Runtime.h [hunk 1]
```diff
@@ -91,6 +91,11 @@ void genTransferSize(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
                      mlir::Value resultBox, mlir::Value sourceBox,
                      mlir::Value moldBox, mlir::Value size);
 
+/// generate runtime call to transfer intrinsic
+void genTransfer(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
+                 mlir::Value resultBox, mlir::Value sourceBox);
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/Runtime.h:96]: Was this `genTransfer` here by accident?

Comment (kiranchandramohan) [reply flang/include/flang/Lower/Runtime.h]: Yeah, that was accidental. Removed and updated comment. Thanks for pointing out.

**Comment by mleair** [comment]
I plan to merge Reshape in a few minutes. The Reshape PR includes a new file called TransformationalRuntime.cpp (and the corresponding header file, TransformationalRuntime.h). We should group all transformationals into these files. It will be consistent with the runtime files transformational.*, and might allow some code sharing through templates, etc.

**Comment by mleair** [comment]
#901 is now merged.

File: flang/lib/Lower/IntrinsicCall.cpp [hunk 2]
```diff
@@ -2552,6 +2558,43 @@ IntrinsicLibrary::genTransfer(mlir::Type resultType,
...
+      Fortran::lower::createTempMutableBox(builder, loc, resultArrayType);
+  auto resultIrBox =
+      Fortran::lower::getMutableIRBox(builder, loc, resultMutableBox);
+  // Call runtime. The runtime is allocating the result.
+  Fortran::lower::genTranspose(builder, loc, resultIrBox, source);
+  // Read result from mutable fir.box and add it to the list of temps to be
+  // finalized by the StatementContext.
+  auto res = Fortran::lower::genMutableBoxRead(builder, loc, resultMutableBox);
+  return res.match(
+      [&](const fir::ArrayBoxValue &box) -> fir::ExtendedValue {
```
Comment (mleair) [inline flang/lib/Lower/IntrinsicCall.cpp:2583]: FYI: PR #901 added a function called readAndAddCleanUp() that can be used to handle this bit of code.

**Comment by kiranchandramohan** [comment]
Thanks @mleair for the review. I have moved the code to TransformationalRuntime.{cpp,h} and readAndAddCleanUp is used now.

File: flang/lib/Lower/Runtime.cpp [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include "../runtime/clock.h"
 #include "../runtime/random.h"
 #include "../runtime/stop.h"
+#include "../runtime/transformational.h"
```
Comment (mleair) [inline flang/lib/Lower/Runtime.cpp:14]: This line is probably not needed anymore, right?

Comment (kiranchandramohan) [reply flang/lib/Lower/Runtime.cpp]: Yes, not needed. Removed.

**Review by jeanPerier — APPROVED** [review]
LGTM, thanks !

**Review by mleair — APPROVED** [review]
Looks good now.

**Pull request merged**